File: libs/statemanager/src/main/kotlin/mono/state/command/mouse/HoverShapeManager.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.state.command.mouse

import mono.graphics.geo.DirectedPoint
import mono.shape.connector.ShapeConnectorUseCase
import mono.shape.shape.AbstractShape
import mono.state.command.CommandEnvironment

/**
 * A short live-time class to manage hover shape.
 * This class is used to avoid searching for hover shape multiple times.
 */
internal class HoverShapeManager private constructor(
    private val searcher: (CommandEnvironment, DirectedPoint) -> AbstractShape?
) {
    private val pointToTargetMap = mutableMapOf<DirectedPoint, AbstractShape?>()

    fun getHoverShape(
        environment: CommandEnvironment,
        point: DirectedPoint
    ): AbstractShape? = pointToTargetMap.getOrSearch(environment, point)

    private fun MutableMap<DirectedPoint, AbstractShape?>.getOrSearch(
        environment: CommandEnvironment,
        point: DirectedPoint
    ): AbstractShape? = getOrPut(point) {
        if (point !in this) {
            searcher.invoke(environment, point)
        } else {
            // This point is already in the map, so we don't need to search again.
            null
        }
    }

    companion object {
        fun forLineConnectHover(): HoverShapeManager = HoverShapeManager { environment, point ->
            ShapeConnectorUseCase.getConnectableShape(
                point,
                environment.getShapes(point.point)
            )
        }
    }
}

File: monosketch-svelte/src/lib/mono/state-manager/command/mouse/hover-shape-manager.ts
import type { Point } from "$libs/graphics-geo/point";
import { getOrNull } from "$libs/sequence";
import { ShapeConnectorUseCase } from "$mono/shape/connector/line-connector";
import type { AbstractShape } from "$mono/shape/shape/abstract-shape";
import type { CommandEnvironment } from "$mono/state-manager/command-environment";

/**
 * A short live-time class to manage hover shape.
 * This class is used to avoid searching for hover shape multiple times.
 */
export class HoverShapeManager {
    private pointToTargetMap: Map<Point, AbstractShape | null> = new Map();

    private constructor(
        private searcher: (environment: CommandEnvironment, point: Point) => AbstractShape | null,
    ) {
    }

    getHoverShape(environment: CommandEnvironment, point: Point): AbstractShape | null {
        return this.pointToTargetMap.get(point) ?? this.searchAndCache(environment, point);
    }

    private searchAndCache(environment: CommandEnvironment, point: Point): AbstractShape | null {
        const shape = this.searcher(environment, point);
        this.pointToTargetMap.set(point, shape);
        return shape;
    }

    resetCache(): void {
        this.pointToTargetMap.clear();
    }

    static forLineConnectHover(): HoverShapeManager {
        return new HoverShapeManager((environment, point) =>
            ShapeConnectorUseCase.getConnectableShape(point, environment.getShapes(point)),
        );
    }

    static forHoverShape(): HoverShapeManager {
        return new HoverShapeManager((environment, point) => {
            const shapes = Array.from(environment.getShapes(point));
            return getOrNull(shapes, shapes.length - 1);
        });
    }
}
